Error with sbt thin client: Invalid build URI

<!-- choose one -->
**Impact**: other

**Tell us about your environment:**
*Chipyard Version:* dev, hash `262c5c8`
*OS:*
* `Linux 3.10.0-957.27.2.el7.x86_64 #1 SMP Mon Jul 29 17:46:05 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux`
* `Linux 5.4.0-67-generic #75-Ubuntu SMP Fri Feb 19 18:03:38 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`

*Other:* <!-- `prior steps taken/documentation followed/...` -->

**What is the current behavior?**

When I try to start up the `sbt` thin client (`ENABLE_SBT_THIN_CLIENT=1 make -C sims/vcs start-sbt-server`), I get an error related to the `sbt.workspace` variable from the `sbt` server process.

**What is the expected behavior?**

The `sbt` server should start up.

**Other information**

This seems related to some of the issues I ran into with setting the `sbt.workspace` variable at the bottom of #555. When I try to start up the `sbt` server, I get this:

```
$ ENABLE_SBT_THIN_CLIENT=1 make -C sims/vcs start-sbt-server
...
cd /path/to/chipyard && java -Xmx8G -Xss8M -XX:MaxPermSize=256M -Djava.io.tmpdir=/path/to/chipyard/.java_tmp -jar /path/to/chipyard/generators/rocket-chip/sbt-launch.jar -Dsbt.sourcemode=true -Dsbt.workspace=/path/to/chipyard/tools --client "exit"
OpenJDK 64-Bit Server VM warning: ignoring option MaxPermSize=256M; support was removed in 8.0
[info] entering *experimental* thin client - BEEP WHIRR
[info] server was not detected. starting an instance
[0m[[0m[0minfo[0m] [0m[0mwelcome to sbt 1.4.4 (Oracle Corporation Java 1.8.0_222)[0m[0J
[0J[0m[[0m[0minfo[0m] [0m[0mloading settings for project chipyardRoot from build.sbt ...[0m[0J
[0J[0m[[0m[0minfo[0m] [0m[0mloading settings for project hardfloat from build.sbt ...[0m[0J
[0J[0m[[0m[0minfo[0m] [0m[0mloading settings for project rocketConfig from build.sbt ...[0m[0J
[0J[0m[[0m[0minfo[0m] [0m[0mloading settings for project firrtl_interpreter from build.sbt ...[0m[0J
[0J[0m[[0m[0minfo[0m] [0m[0mloading settings for project treadle from build.sbt ...[0m[0J
[0J[0m[[0m[0minfo[0m] [0m[0mloading settings for project chisel_testers from build.sbt ...[0m[0J
[0J[0m[[0m[0minfo[0m] [0m[0mloading settings for project testchipip from build.sbt ...[0m[0J
[0J[0m[[0m[0minfo[0m] [0m[0mloading settings for project icenet from build.sbt ...[0m[0J
[0J[0m[[0m[0minfo[0m] [0m[0mloading settings for project hwacha from build.sbt ...[0m[0J
[0J[0m[[0m[0minfo[0m] [0m[0mloading settings for project boom from build.sbt ...[0m[0J
[0J[0m[[0m[0minfo[0m] [0m[0mloading settings for project cva6 from build.sbt ...[0m[0J
[0J[0m[[0m[0minfo[0m] [0m[0mloading settings for project sodor from build.sbt ...[0m[0J
[0J[0m[[0m[0minfo[0m] [0m[0mloading settings for project gemmini from build.sbt ...[0m[0J
[0J[0m[[0m[0minfo[0m] [0m[0mloading settings for project mdf from build.sbt ...[0m[0J
[0J[0m[[0m[0minfo[0m] [0m[0mloading settings for project barstoolsMacros from build.sbt ...[0m[0J
[0J[0m[[0m[0minfo[0m] [0m[0mloading settings for project niobium_blocks from build.sbt ...[0m[0J
[0J[0m[[0m[0minfo[0m] [0m[0mloading settings for project sim-build from plugins.sbt ...[0m[0J
[0J[0m[[0m[0minfo[0m] [0m[0mloading project definition from /path/to/chipyard/sims/firesim/sim/project[0m[0J
[0J[0m[[0m[0minfo[0m] [0m[0mloading settings for project firesim from build.sbt ...[0m[0J
[0J[0m[[0m[0minfo[0m] [0m[0mloading settings for project targetutils from build.sbt ...[0m[0J
[0J[0m[[0m[0minfo[0m] [0m[0mloading settings for project midas from build.sbt ...[0m[0J
[0Jjava.lang.RuntimeException: Invalid build URI (no handler available): file:///home/users/tynan/workspace/chisel3/
... (stacktrace follows)
```

So, it's looking in my home directory for `chisel3`, rather than in `/path/to/chipyard/tools`. If I had to guess, that's because while the `sbt` thin client gets `-Dsbt.workspace` set at the command-line, the `sbt` server instance doesn't get that option (or any of the other options). However, if I create a `.sbtopts` file that looks like this:

```
$ cat .sbtopts
-Dsbt.sourcemode=true
-Dsbt.workspace=$PWD/tools
```

... then that `start-sbt-server` command succeeds, and I can run `ENABLE_SBT_THIN_CLIENT=1 make -C sims/vcs verilog`, which interacts with the `sbt` server.

Should we be using a `.sbtopts` file rather than passing those arguments around the `SBT_OPTS` variable in `variables.mk`? Or is there some way to pass arguments to the `sbt` server process? I'm actually also running into a bunch of Java garbage collection performance issues with the `sbt` server, which makes me think it's also not getting the `-Xmx8G -Xss8M -XX:MaxPermSize=256M` settings.

I got the same two errors of @tymcauley and @shi27feng in #555. I did what @tymcauley wrote and now I can run `make launch-sbt` and everything goes smooth. I also added the `.sbtopts` file and I can run `sbt` from the root direcotry without problems.

I have an other problem though. If I run in any directory `make launch-sbt` there are no problems. However, if I load the sbt changes from IntelliJ Idea this error shows up:
```
[error] Invalid build URI (no handler available): file:/home/davide/Projects/chipyard/$PWD/tools/chisel3/
```

Moreover, if I try, inside the folder `fpga`, running `make SUB_PROJECT=vcu118 bitstream`, the exact same error typology of @shi27feng shows up. More precisely, the error is the subsequent: 
```
make: *** [/home/davide/Projects/chipyard/common.mk:86: /home/davide/Projects/chipyard/fpga/generated-src/chipyard.fpga.vcu118.VCU118FPGATestHarness.RocketVCU118Config/sim_files.f] Error 1
```
I am simply trying to see how the bitstream is generated for the board above. I am new to Chipyard.

I'm currently on the master branch (latest release), running on an Ubuntu 20.04 machine. 

Thank you very much in advance.

--- a/.sbtopts
+++ b/.sbtopts
@@ -1,0 +1,2 @@
+-Dsbt.sourcemode=true
+-Dsbt.workspace=$PWD/tools

--- a/common.mk
+++ b/common.mk
@@ -247,12 +247,17 @@ SBT_COMMAND ?= shell
 launch-sbt:
 	cd $(base_dir) && $(SBT_NON_THIN) "$(SBT_COMMAND)"
 
+check-thin-client:
+ifeq (,$(ENABLE_SBT_THIN_CLIENT))
+	$(error ENABLE_SBT_THIN_CLIENT not set.)
+endif
+
 .PHONY: shutdown-sbt-server
-shutdown-sbt-server:
+shutdown-sbt-server: check-thin-client
 	cd $(base_dir) && $(SBT) "shutdown"
 
 .PHONY: start-sbt-server
-start-sbt-server:
+start-sbt-server: check-thin-client
 	cd $(base_dir) && $(SBT) "exit"
 
 #########################################################################################

--- a/common.mk
+++ b/common.mk
@@ -18,7 +18,7 @@ HELP_COMPILATION_VARIABLES += \
 "   EXTRA_SIM_LDFLAGS      = additional LDFLAGS for building simulators" \
 "   EXTRA_SIM_SOURCES      = additional simulation sources needed for simulator" \
 "   EXTRA_SIM_REQS         = additional make requirements to build the simulator" \
-"   ENABLE_SBT_THIN_CLIENT = if set, use sbt's experimental thin client"
+"   ENABLE_SBT_THIN_CLIENT = if set, use sbt's experimental thin client (works best with sbtn or sbt script)"
 
 EXTRA_GENERATOR_REQS ?=
 EXTRA_SIM_CXXFLAGS   ?=

--- a/common.mk
+++ b/common.mk
@@ -162,16 +162,21 @@ verilog: $(sim_vsrcs)
 #########################################################################################
 .PHONY: run-binary run-binary-fast run-binary-debug run-fast
 
+check-binary:
+ifndef BINARY
+	$(error BINARY variable is not set. Set it to the simulation binary)
+endif
+
 # run normal binary with hardware-logged insn dissassembly
-run-binary: $(output_dir) $(sim)
+run-binary: $(output_dir) $(sim) $(check-binary)
 	(set -o pipefail && $(NUMA_PREFIX) $(sim) $(PERMISSIVE_ON) $(SIM_FLAGS) $(EXTRA_SIM_FLAGS) $(SEED_FLAG) $(VERBOSE_FLAGS) $(PERMISSIVE_OFF) $(BINARY) </dev/null 2> >(spike-dasm > $(sim_out_name).out) | tee $(sim_out_name).log)
 
 # run simulator as fast as possible (no insn disassembly)
-run-binary-fast: $(output_dir) $(sim)
+run-binary-fast: $(output_dir) $(sim) $(check-binary)
 	(set -o pipefail && $(NUMA_PREFIX) $(sim) $(PERMISSIVE_ON) $(SIM_FLAGS) $(EXTRA_SIM_FLAGS) $(SEED_FLAG) $(PERMISSIVE_OFF) $(BINARY) </dev/null | tee $(sim_out_name).log)
 
 # run simulator with as much debug info as possible
-run-binary-debug: $(output_dir) $(sim_debug)
+run-binary-debug: $(output_dir) $(sim_debug) $(check-binary)
 	(set -o pipefail && $(NUMA_PREFIX) $(sim_debug) $(PERMISSIVE_ON) $(SIM_FLAGS) $(EXTRA_SIM_FLAGS) $(SEED_FLAG) $(VERBOSE_FLAGS) $(WAVEFORM_FLAG) $(PERMISSIVE_OFF) $(BINARY) </dev/null 2> >(spike-dasm > $(sim_out_name).out) | tee $(sim_out_name).log)
 
 run-fast: run-asm-tests-fast run-bmark-tests-fast
@@ -182,16 +187,19 @@ run-fast: run-asm-tests-fast run-bmark-tests-fast
 $(binary_hex): $(output_dir) $(BINARY)
 	$(base_dir)/scripts/smartelf2hex.sh $(BINARY) > $(binary_hex)
 
+run-binary-hex: $(check-binary)
 run-binary-hex: $(output_dir) $(sim) $(binary_hex)
 run-binary-hex: run-binary
 run-binary-hex: override LOADMEM_ADDR = 80000000
 run-binary-hex: override LOADMEM = $(binary_hex)
 run-binary-hex: override SIM_FLAGS += +loadmem=$(LOADMEM) +loadmem_addr=$(LOADMEM_ADDR)
+run-binary-debug-hex: $(check-binary)
 run-binary-debug-hex: $(output_dir) $(sim) $(binary_hex)
 run-binary-debug-hex: run-binary-debug
 run-binary-debug-hex: override LOADMEM_ADDR = 80000000
 run-binary-debug-hex: override LOADMEM = $(binary_hex)
 run-binary-debug-hex: override SIM_FLAGS += +loadmem=$(LOADMEM) +loadmem_addr=$(LOADMEM_ADDR)
+run-binary-fast-hex: $(check-binary)
 run-binary-fast-hex: $(output_dir) $(sim) $(binary_hex)
 run-binary-fast-hex: run-binary-fast
 run-binary-fast-hex: override LOADMEM_ADDR = 80000000

--- a/common.mk
+++ b/common.mk
@@ -163,20 +163,20 @@ verilog: $(sim_vsrcs)
 .PHONY: run-binary run-binary-fast run-binary-debug run-fast
 
 check-binary:
-ifndef BINARY
+ifeq (,$(BINARY))
 	$(error BINARY variable is not set. Set it to the simulation binary)
 endif
 
 # run normal binary with hardware-logged insn dissassembly
-run-binary: $(output_dir) $(sim) $(check-binary)
+run-binary: $(output_dir) $(sim) check-binary
 	(set -o pipefail && $(NUMA_PREFIX) $(sim) $(PERMISSIVE_ON) $(SIM_FLAGS) $(EXTRA_SIM_FLAGS) $(SEED_FLAG) $(VERBOSE_FLAGS) $(PERMISSIVE_OFF) $(BINARY) </dev/null 2> >(spike-dasm > $(sim_out_name).out) | tee $(sim_out_name).log)
 
 # run simulator as fast as possible (no insn disassembly)
-run-binary-fast: $(output_dir) $(sim) $(check-binary)
+run-binary-fast: $(output_dir) $(sim) check-binary
 	(set -o pipefail && $(NUMA_PREFIX) $(sim) $(PERMISSIVE_ON) $(SIM_FLAGS) $(EXTRA_SIM_FLAGS) $(SEED_FLAG) $(PERMISSIVE_OFF) $(BINARY) </dev/null | tee $(sim_out_name).log)
 
 # run simulator with as much debug info as possible
-run-binary-debug: $(output_dir) $(sim_debug) $(check-binary)
+run-binary-debug: $(output_dir) $(sim_debug) check-binary
 	(set -o pipefail && $(NUMA_PREFIX) $(sim_debug) $(PERMISSIVE_ON) $(SIM_FLAGS) $(EXTRA_SIM_FLAGS) $(SEED_FLAG) $(VERBOSE_FLAGS) $(WAVEFORM_FLAG) $(PERMISSIVE_OFF) $(BINARY) </dev/null 2> >(spike-dasm > $(sim_out_name).out) | tee $(sim_out_name).log)
 
 run-fast: run-asm-tests-fast run-bmark-tests-fast
@@ -187,19 +187,19 @@ run-fast: run-asm-tests-fast run-bmark-tests-fast
 $(binary_hex): $(output_dir) $(BINARY)
 	$(base_dir)/scripts/smartelf2hex.sh $(BINARY) > $(binary_hex)
 
-run-binary-hex: $(check-binary)
+run-binary-hex: check-binary
 run-binary-hex: $(output_dir) $(sim) $(binary_hex)
 run-binary-hex: run-binary
 run-binary-hex: override LOADMEM_ADDR = 80000000
 run-binary-hex: override LOADMEM = $(binary_hex)
 run-binary-hex: override SIM_FLAGS += +loadmem=$(LOADMEM) +loadmem_addr=$(LOADMEM_ADDR)
-run-binary-debug-hex: $(check-binary)
+run-binary-debug-hex: check-binary
 run-binary-debug-hex: $(output_dir) $(sim) $(binary_hex)
 run-binary-debug-hex: run-binary-debug
 run-binary-debug-hex: override LOADMEM_ADDR = 80000000
 run-binary-debug-hex: override LOADMEM = $(binary_hex)
 run-binary-debug-hex: override SIM_FLAGS += +loadmem=$(LOADMEM) +loadmem_addr=$(LOADMEM_ADDR)
-run-binary-fast-hex: $(check-binary)
+run-binary-fast-hex: check-binary
 run-binary-fast-hex: $(output_dir) $(sim) $(binary_hex)
 run-binary-fast-hex: run-binary-fast
 run-binary-fast-hex: override LOADMEM_ADDR = 80000000

--- a/project/build.properties
+++ b/project/build.properties
@@ -1,1 +1,1 @@
-sbt.version=1.4.4
+sbt.version=1.4.9

--- a/variables.mk
+++ b/variables.mk
@@ -152,7 +152,10 @@ JAVA_OPTS ?= -Xmx$(JAVA_HEAP_SIZE) -Xss8M -XX:MaxPermSize=256M -Djava.io.tmpdir=
 # default sbt launch command
 #########################################################################################
 # by default build chisel3/firrtl and other subprojects from source
-override SBT_OPTS += -Dsbt.sourcemode=true -Dsbt.workspace=$(base_dir)/tools
+SBT_OPTS_FILE := $(base_dir)/.sbtopts
+ifneq (,$(wildcard $(SBT_OPTS_FILE)))
+override SBT_OPTS += $(subst $$PWD,$(base_dir),$(shell cat $(SBT_OPTS_FILE)))
+endif
 
 SCALA_BUILDTOOL_DEPS = $(SBT_SOURCES)
 

--- a/variables.mk
+++ b/variables.mk
@@ -164,6 +164,7 @@ SBT_THIN_CLIENT_TIMESTAMP = $(base_dir)/project/target/active.json
 ifdef ENABLE_SBT_THIN_CLIENT
 override SCALA_BUILDTOOL_DEPS += $(SBT_THIN_CLIENT_TIMESTAMP)
 # enabling speeds up sbt loading
+# use with sbt script or sbtn to bypass error code issues
 SBT_CLIENT_FLAG = --client
 endif
 

